### ISSUE TITLE
Missing error details in nested sub-models validations

### ISSUE THREAD

**Issue by joseandrespg** [sylvainpolletvillard/ObjectModel#38]
Hi, 

Following https://github.com/sylvainpolletvillard/ObjectModel/issues/37

We miss extra information when an error happen inside a sub-model validation. 

Given the example:

```javascript
var Model = require("objectmodel");

function printErrors(errors) {
    console.log("Caught these errors:");
    errors.forEach(function(error){ console.log(error); });
}

var CountryCode = new Model(String)
.assert(function (a) {
    return a === "GB";
}, "Country must be GB");

var Address = new Model({
    city: String,
    address: String,
    country: CountryCode
});

var Order = new Model({
    sku: String,
    addresses: Model.Array(Address)
});

// Examples
var gbAddress = { city: "London", address: "123 Fake St", country: "GB"}; // valid address
var frAddress = { city: "Paris", address: 0, country: "FR" }; // invalid address and country

var gbOrder = { sku: "ABC123", addresses: [gbAddress] };
var frOrder = { sku: "ABC123", addresses: [frAddress] };

Order.validate(gbOrder, printErrors); // no errors
Order.validate(frOrder, printErrors); // throw errors
```

The actual console output is:

```
Caught these errors:
{ expected: [Function: String],
  received: 0,
  path: 'addresses[0].address',
  message: 'expecting addresses[0].address to be String, got Number 0' }

{ message: 'assertion "Country must be GB" returned false for value "FR"' }
```

The expected output would be something like:

```
Caught these errors:
{ expected: [Function: String],
  received: 0,
  path: 'addresses[0].address',
  message: 'expecting addresses[0].address to be String, got Number 0' }

{ expected: [Function: String],
  received: 'FR',
  path: 'addresses[0].country',
  message: 'Country must be GB' }
```

That will give us the ability to detect the exact path of an error.

Thanks!

**Comment by sylvainpolletvillard** [comment]
Assertion errors won't provide an `expected` parameter because this parameter points to the model definition which is valid here. But I can add the `received` parameter. 

Also there is work in progress to provide `path` information inside complex ObjectModel/ArrayModel nesting.

**Comment by sylvainpolletvillard** [comment]
I guess I can set the `expected` parameter to the assertion description (or source code if no description is provided).

![image](https://cloud.githubusercontent.com/assets/566536/21941192/e21e0034-d9c7-11e6-9c33-8e39965a581b.png)

What do you think ?

**Comment by joseandrespg** [comment]
I would prefer to keep `expected` always as a function (source code). I would be more predictable.

I think having the assertion's text in the message should be enough.

Another additional solution would be add internally the text as a property in the function:

```javascript
// Source code
var CountryCode = new Model(String)
.assert(function assertionFunction(a) {
    return a === "GB";
}, "Country must be GB");

// Internally 
// Copy the assert message to a new `message` property. It will be available in the error object
assertionFunction.message = "Country must be GB"; 
```

It shouldn't be dangerous...

What are your thoughts about this?

**Comment by sylvainpolletvillard** [comment]
In previous versions, descriptions were added directly to the assertion functions as property `_description`. But there was a problem when using the same assertion on different models with different descriptions. So the new behaviour is to add a `_onFail` property that has description in its closure.

`expected` is not always a function, currently it is used only on definition checking and reflects the model definition part. So it can be a constructor, a simple value or whatever you put in your model definition. Nothing really predictable here. But I'm okay to put a reference to the assertion function in the `expected` property ; `message` should be good enough for most cases and the descriptions can be easily extracted.

**Comment by sylvainpolletvillard** [comment]
fixed in v2.6 just released now :+1:

**Comment by joseandrespg** [comment]
Thanks @sylvainpolletvillard ! :clap:
